### ISSUE TITLE
 [BUG-General]  OSError: dlopen() failed to load a library: pango / pango-1 / pango-1.0 / pango-1.0-0

### ISSUE THREAD

**Issue by glipR** [ManimCommunity/manim#695]
**Describe the bug**
After updating to the latest version of manim, I ran into the same issue as #689 . I tried the fixes located at https://docs.manim.community/en/latest/installation/troubleshooting.html#modulenotfounderror-no-module-named-generated-ffi

I don't have a ProgramFiles/Intel/Wifi folder. my zlib1.dll comes from MingW. I replaced the MingW zlib1.dll with the pango provided file and the `python pangocairocffi/ffi_build.py` command still fails (The other two pass) complaining it cannot load a library `pango / pango-1/...`.

**To Reproduce**
I'd originally installed with chocolatey. Today I tried to reinstall with chocolatey, and ran into the pangocairocffi error.
I've also tried downloading the pango binaries separately and adding to path.
Additionally, I've used pip install manimce in a virtualenv.

**Expected behavior**
The pango build script succeeds.

**Logs**
<details><summary>The massive chocolatey logs</summary>

https://pastebin.com/xaHrXUSm
</details>

**System Specifications**

<details><summary>System Details</summary>

- OS: Windows 10 10.0.19041
- RAM: 8GB
- Python version 3.8.6 (venv)
- Installed modules:
```
Package           Version
----------------- -------
cairocffi         1.1.0
cffi              1.14.3
colorama          0.4.4
colour            0.1.5
commonmark        0.9.1
grpcio            1.33.2
grpcio-tools      1.33.2
manimce           0.1.0
numpy             1.19.3
pangocairocffi    0.3.2
pangocffi         0.6.0
pathtools         0.1.2
Pillow            8.0.1
pip               20.2.4
progressbar       2.5
protobuf          3.13.0
pycairo           1.20.0
pycparser         2.20
pydub             0.24.1
Pygments          2.7.2
rich              9.2.0
scipy             1.5.4
setuptools        50.3.2
six               1.15.0
tqdm              4.51.0
typing-extensions 3.7.4.3
watchdog          0.10.3
wheel             0.35.1
```

</details>

**Comment by behackl** [comment]
Thank you for your report! This is difficult to debug from the information provided, but let's see: are you absolutely sure that your version of `zlib1.dll` is now provided via pango?

Otherwise: is there any other information you can provide so that we can figure out why precisely loading pango fails? It seems to me like your PATH contains another patched version of some library that is picked up over the version provided by pango -- a tedious, but probably successful approach would be to check whether each library shipped with Pango is actually also picked up first on your PATH; you can use `WHERE library.dll` in a cmd or so.

Also pinging @naveen521kk -- more ideas what could be at fault here are welcome.

**Comment by glipR** [comment]
Thanks for the ideas. I'd already checked `which zlib1.dll` points to the pango provided file, although I haven't done it for the rest, I can try that later.

Worth noting that before I updated manim I was running into issues with my virtual environment https://stackoverflow.com/questions/38132755/importerror-no-module-named-encodings then reinstalled manimce on a new venv.

**Comment by naveen521kk** [comment]
Windows is weird. The DLL once loaded to memory will reuse that even though the content is different, and eventhough is on top of the path. So, maybe any other app loaded it before and is kept mixed up? 
I know this is frustrating to say.

#### How to fix it?
Sometimes simply restarting your computer will fix it but sometimes not. If that doesn't help I will help you out with a small workaround by hacking the loaded dll using ctypes. But that would be frustrating to do each time you run manim. So, here is my workaround for it.

- First, start a python interative shell, in your case it `py -3.8`. Assuming you have those at first in the path, type in the following.
 ```py
 from ctypes.util import find_library
 a=find_library("libpango-1.0-0")
 import ctypes
 ctypes.CDLL(a)
 ```
Now you should be doing the above thing replacing `libpango-1.0-0` with `libpangocairo-1.0-0.dll`,`libcairo-2.dll`,`zlib1.dll`,`libgobject-2.0-0.dll` .
It should then probally work when you run manim next.(I haven't tried it though).

- One thing we can possiblly do is build it ourselves rather then making use of these from msys2, I had build it for one of the other projects, you could try downloading [this](http://r.syrusdark.website/download-pango) and try out the same. Maybe I should update it in the other repo?

**Comment by naveen521kk** [comment]
Also, I more towards making `pango` optional. What do you think about this @ManimCommunity/core ?

**Comment by glipR** [comment]
Ok, I was able to figure out my issue thanks to your workaround. It turns out it was referencing my already existing version of pango that came with GTK. (`ctypes.CDLL(a)` tries to open `C:\GTK\bin\libpango-1.0-0.dll`, which then complains it is not a valid Win32 application.)

I was able to fix it by removing the GTK folder, running the 4 lines of code above (just `libpango-1.0-0` was all I needed), then forcing a new build by cd-ing into my sitepackages and running `python pangocairocffi/ffi_build.py`. So if anyone else runs into the same issue with the choco install I would try `find_library("libpango-1.0-0")` and ensure that you are trying to load the `dll` from `C:\tools\pango`.

Thank you for the information @naveen521kk ! I'm happy to close this issue but if you'd rather keep it open for further discussion regarding your last comment then thats cool too.

**Comment by leotrs** [comment]
@naveen521kk make PangoText optional and make the default CairoText? Or do you mean to make Text optional altogether, regardless of the backend, and encourage Tex instead?

**Comment by naveen521kk** [comment]
I meant making `Text` to default to Pango, but if there is any error, like above just dont raise an error, else issue a warning and then use `Cairo` backend. 
If pango, is available it is used, if not, Cairo is used.

**Comment by leotrs** [comment]
I see. I'm not opposed to having Cairo as a fallback, but I think development of the Pango backend should still continue.

Also, I'd much rather this not be automatic. I'd rather raise all those errors and ask the user to explicitly change from Text (or PangoText) to CairoText. The reason behind this is that if we don't see those errors, we will never have a complete implementation of Pango.

Let's see what others think first.

**Comment by huguesdevimeux** [comment]
To be honest I don't really mind. 
Just one thought tho : We have to be sure that stuff that use Text (such as CodeMobject) does not depend specifically on one. That's not yet the case IIRC but could be in the future specifically regarging CodeMobject. 

But, why not refactoring everything and doing a abstract class `Text`, overriden by `CairoText` and `PangoText` (name are just placeholders)

**Comment by MulleDK19** [comment]
These are the issues I ran into, and how I fixed them.

1. Chocolatey installs Python 3.8 without PIP, which causes it to fail early. Installing Python 3.8 with PIP manually first fixes that issue.
2. Failing to load pango. I fixed this by moving c:\tools\pango to the top of PATH (right after Python), then running refreshenv and forcing a reinstall of manimce using `choco install mcanimce --force`. This caused the right libraries to be loaded.
